Azure PD (Managed/Blob)

khenidak opened the pull request:
This is exactly the same code as this [PR](https://github.com/kubernetes/kubernetes/pull/41950). It has a clean set of generated items. We created a separate PR to accelerate the accept/merge the PR

CC @colemickens 
CC @brendandburns 

**What this PR does / why we need it**:

1. Adds K8S support for Azure Managed Disks. 
2. Adds support for dedicated blob disks (1:1 to storage account) in addition to shared blob disks (n:1 to storage account). 
3. Automatically manages the underlying storage accounts. New storage accounts are created at 50% utilization. Max is 100 disks, 60 disks per storage account.    
2. Addresses the current issues with Blob Disks:
..* Significantly faster attach process. Disks are now usually available for pods on nodes under 30 sec if formatted, under a min if not formatted. 
..* Adds support to move disks between nodes.
..* Adds consistent attach/detach behavior, checks if the disk is leased/attached on a different node before attempting to attach to target nodes.
..* Fixes a random hang behavior on Azure VMs during mount/format (for both blob + managed disks).
..* Fixes a potential conflict by avoiding the use of disk names for mount paths. The new plugin uses hashed disk uri for mount path.  

The existing AzureDisk is used as is. Additional "kind" property was added  allowing the user to decide if the pd will be shared, dedicated or managed (Azure Managed Disks are used).

Due to the change in mounting paths, existing PDs need to be recreated as PV or PVCs on the new plugin.

roberthbailey commented on pkg/cloudprovider/providers/azure/azure_file.go:
```diff
@@ -23,6 +23,10 @@ import (
 	azs "github.com/Azure/azure-sdk-for-go/storage"
 )
 
+const (
```
Please put this onto a single line instead of using a block since there is only one constant.

roberthbailey reviewed (commented):
I've only reviewed the code in the cloudprovider directory. The majority of the changes are in the volume/azure_dd package, so I'll defer to @ brendandburns and @rootfs to review and approve those changes.

roberthbailey commented:
/approve

rootfs commented:
@k8s-bot ok to test

rootfs commented:
@khenidak This is a large refactoring, hard to review.

Please reuse existing code and interface and not shuffle/copy them around. If there existing interface not working for you, address the limitation in the interface.

thockin commented:
/approve

This is not a code review

khenidak commented:
@rootfs 

1. Moved the disk controllers to Azure Cloud as requested.
2. re REST vs SDK - We will move to the SDK once we test against the specific requirements for plugin
3. Added few unit tests where possible.

rootfs commented on pkg/volume/azure_dd/common.go:
move ioHandler and below out of this file - they are going to be consolidated with other volume plugins.

rootfs commented on pkg/volume/azure_dd/common.go:
this is mostly identical (except the order of command) with `/pkg/util/mount` but bypass mounter and directly call `exec.New` to obtain a runner. This is problematic as we are refactoring mounters. 

If there is an issue with default mounter, refactor at `/pkg/util/mount`, but don't bypass mounter interface.

rootfs commented on pkg/volume/azure_dd/mounter.go:
```diff
@@ -0,0 +1,179 @@
+/*
```
follow the file name convention: the file that implement the plugin is named with the plugin's name.

khenidak replied:
the file is named azure_dd.go, as for mount and attacher i didn't find a consistent naming concision across the plugins. I can prefix all files with azure_dd if that will help.

rootfs replied:
filename prefixed with azure_ for log parsing. 
For new files, copyright year stay with 2017

rootfs commented on pkg/volume/azure_dd/vhd_util.go:
```diff
@@ -1,145 +0,0 @@
-/*
```
keep it as we are going to refactor disk utilities among plugins.

khenidak replied:
this has moved into the common file of the disk controllers. Since this only needed by Blob Disk + Managed Disk (and will *always* be, since any block device based disks will always use AzureDiskVolumeSource type with the addition of Kind parameter) also they have nothing to do with "vhd" per-se

rootfs commented on pkg/volume/azure_dd/mounter.go:
```diff
@@ -0,0 +1,179 @@
+var _ volume.Unmounter = &azureDiskUnmounter{}
+var _ volume.Mounter = &azureDiskMounter{}
+
+func (m *azureDiskMounter) GetAttributes() volume.Attributes {
```
Missed `Managed`

khenidak replied:
added

rootfs commented on pkg/volume/azure_dd/mounter.go:
```diff
@@ -0,0 +1,179 @@
+	}
+
+	// we didn't fail
+	//TODO: do we need to do this? the bind-mount performs ro mount
```
this is for ownership change.

khenidak replied:
removed the comment

rootfs commented:
@khenidak there are many regressions when you remove existing `azure_dd.go` and decide to withdraw support for provsioning parameters.

In existing `https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/azure_dd/azure_dd.go`, you'll find newly added functions such as `SupportsMountOption` and `SupportsBulkVolumeVerification`. Even in functions you did copy from there, the function signature changes.

As we are approaching code freeze, extensive refactoring must be carefully evaluated.

rootfs commented:
cc @jdumars

khenidak commented on pkg/volume/azure_dd/common.go:
```diff
@@ -0,0 +1,322 @@
+	podUID     types.UID
+}
+
+var supportedCachingModes = sets.NewString(
```
changed

khenidak commented on pkg/volume/azure_dd/common.go:
by the time this function is called, validation has already occurred at the time of Init() call on the plugin

khenidak commented on pkg/volume/azure_dd/common.go:
attached.go:213

khenidak commented on pkg/volume/azure_dd/common.go:
azure_dd.go:150
azure_dd.go:175
azure_dd.go:186 

mainly a short cut to create a model that carries disk metadata around.

khenidak commented on pkg/volume/azure_dd/common.go:
Please help understand more move it where?

khenidak commented on pkg/volume/azure_dd/common.go:
as discussed, there is an issue that hangs the format/mount sequence as its is in K8s (it takes 4 mins+ to execute, which is basically a hang). It waits on IO (process in D sate) on Azure. Until we identify exactly what is wrong (either in k8s) & until a fix is applied and taste (either on Azure/k8s or both) we will have to go with this approach since this approach consistently works with predictable format/mount time.

khenidak commented on pkg/volume/azure_dd/common.go:
discussed above.

khenidak commented on pkg/cloudprovider/providers/azure/blobDiskController.go:
```diff
@@ -0,0 +1,1036 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
actually it is 2016 in all the files i have checked. Wouldn't that break Bazel if changed?

brendandburns replied:
No, 2017 for all new files should work.

khenidak commented on pkg/volume/azure_dd/provisioner.go:
```diff
@@ -0,0 +1,170 @@
+		return nil, fmt.Errorf("azureDisk - claim.Spec.Selector is not supported for dynamic provisioning on Azure disk")
+	}
+
+	if len(p.options.PVC.Spec.AccessModes) > 1 {
```
Yes. It is not elegant. Verbosity meant to help the user understand what went wrong with every case instead of cryptic error messages.

rootfs replied:
what cryptic error messages?

khenidak commented on pkg/volume/azure_dd/provisioner.go:
because sku is "arm" label not end user label, all Azure documentation and portal refer to it "storage account type". Only when you interact with arm templates you deal with sku. Hence, This should use "storage account type" to avoid confusion, but since there is no risk (as described below) in leaving sku parameter it is left for this version until we remove it from all samples then retire it

khenidak commented on pkg/volume/azure_dd/provisioner.go:
Users does not need to provide location, location is picked up from the cloud provider and disks are created in the same resource group. The current implementation does validate location = current (AFAIK) hence is subject to situation where users create disks in different location than the cluster. PVCs and PVs provisioned already will have no impact with this change. only new PVC/PVs

khenidak commented on pkg/volume/azure_dd/provisioner.go:
because the user has 2 options when doing blob disk, either PVC (dedicated or shared) using a set of storage accounts managed by the plugin or PV in this case the user provides diskuri which includes storage account

khenidak commented:
Support bulk verification and support mount options are there in this PR for a while now including passing the mount options to the mounter during mount.  as for signature change if you are referring to referring to UnixxxxID type defs then these have also been included.

I went back and verified that this PR includes all the "bulk" changes that happened in volume. including one related to skipping TearDownAt call if the path does not exist (just updated the PR with it).

rootfs commented on pkg/cloudprovider/providers/azure/controllerCommon.go:
```diff
@@ -0,0 +1,383 @@
+		return attached, lun, err
+	}
+
+	fragment := vmData.(map[string]interface{})
```
This is not safe, it will panic on unexpected `vm` stream. Run unit test below and you'll get a panic.

```go
 	vm := []byte("foo")
	if err := json.Unmarshal(vm, &vmData); err != nil {
               return 
	}
        // panic happens!
  	fragment := vmData.(map[string]interface{})
```

andyzhangx replied:
thanks for the check, I have used below code to avoid such conversion issue:
	fragment, ok := vmData.(map[string]interface{})
	if !ok {
		return attached, lun, fmt.Errorf("convert vmData to map error")
	}
Also do the checks for the proceeding code.

andyzhangx replied:
I have movoed all the error handling code of json format in func: ExtractVMData, ExtractDiskData in azure_util.go

rootfs commented on pkg/cloudprovider/providers/azure/controllerCommon.go:
```diff
@@ -0,0 +1,383 @@
+	diskIdTemplate       string = "/subscriptions/%s/resourcegroups/%s/providers/microsoft.compute/disks/%s"
+	defaultDataDiskCount int    = 16 // which will allow you to work with most medium size VMs (if not found in map)
+
+	vhdBlobUriTemplate             = "https://%s.blob.core.windows.net/%s/%s"
```
This is not accurate. You have to choose among `core.windows.net`, `core.usgovcloudapi.net`, etc....

andyzhangx replied:
thanks for the hint, I have used az.Environment.StorageEndpointSuffix

rootfs commented on pkg/cloudprovider/providers/azure/controllerCommon.go:
```diff
@@ -0,0 +1,383 @@
+
+const (
+	aadTokenEndPointPath string = "%s/oauth2/token/"
+	apiversion           string = "2016-04-30-preview"
```
This apiversion is not supported in every azure region
https://github.com/Azure/azure-xplat-cli/issues/3513

andyzhangx replied:
For ARM, The supported api-versions are '2016-04-30-preview, 2017-03-30' in global azure. So for other regions, like azure china, just wait for the api version update or do customerization if want to use it ASAP.

rootfs commented on pkg/volume/azure_dd/common.go:
keep vhd_util.go file and its unit test.

rootfs commented:
@khenidak address these concerns and ping me or other members from @kubernetes/sig-storage-pr-reviews when you are done

## Make CI test pass.
None of the CI tests have passed yet. Run `hach/verify-all.sh`

## Unit test new functions
Especially those in raw HTTP, there are many uncaptured errors.

## Refactoring:
- make sure existing methods in [azure_dd.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/azure_dd/azure_dd.go) are not lost 
- Address mount utility related in issues, if any, in `pkg/util/mount`
- Keep vhd_util.go and its unit test

## Compatibility
- We need to support `location` parameter in Provision(). Provisioner doesn't always run in the same region as compute instance, 
- Don't withdraw support for parameters without a proper notice or documentation. 

## Raw HTTP:
- If you continue raw HTTP, handle errors and malformed json.
- Make sure the API version works for all azure regions that are currently supported in existing code.
- Don't use hardcoded azure endpoints in raw HTTP, they don't work in every azure region. 
I am not an expert on Azure sdk, cc @ahmetb @colemickens 

## Address review comments from @brendandburns in #41950

cc @jdumars

rootfs commented on pkg/cloudprovider/providers/azure/blobDiskController.go:
cc @colemickens 

As advised by Joseph Romeo by email: "Avoid looking at VHD blob leases to decide a disk is attached or detached.  This can lead you to unexpected failures or worse-case get into a bad state."

rootfs commented on pkg/volume/azure_dd/attacher.go:
```diff
@@ -24,191 +24,168 @@ import (
+	cachingMode := string(*volumeSource.CachingMode)
+	managed := (*volumeSource.Kind == v1.AzureManagedDisk)
+	if managed {
+		lun, err = a.plugin.diskController.AttachManagedDisk(string(nodeName), volumeSource.DataDiskURI, cachingMode)
```
nodeName != InstanceID
refer to https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/azure_dd/attacher.go#L76

andyzhangx replied:
hi, we use nodeName to get the vm object, see in AttachManagedDisk func:
vm, err := c.common.getArmVM(nodeName)

rootfs commented on pkg/cloudprovider/providers/azure/controllerCommon.go:
```diff
@@ -0,0 +1,383 @@
+}
+
+// finds an empty based on VM size and current attached disks
+func findEmptyLun(vmSize string, dataDisks []interface{}) (int, error) {
```
How to deal with race condition?

andyzhangx replied:
I have used lock in this func

rootfs commented on pkg/volume/azure_dd/provisioner.go:
see https://github.com/kubernetes/kubernetes/issues/37801

@jsafrane

brendandburns commented on pkg/volume/azure_dd/common.go:
I'd really rather not fix this problem in this PR. We can TODO it, but this PR has a bunch of fixes I'd like to see merged. Can we address the refactor/merge into pkg/util/mount in a separate PR?

saad-ali commented:
@rootfs @khenidak FYI, code freeze for 1.7 is at 6 PM PT today (about 1 hour away). For this to get into 1.7.0, it must be LGTM'd before then.

mikedanese commented:
Please rebase instead of including merge commits in your PR.

nmakhotkin commented:
Hi! 

Recently I tried to create my own k8s cluster on azure (v1.6.2) not using azure container engine but using my own install tools (some code written in Go and custom-data scripts in bash). Cluster runs successfully except mounting and attaching volumes: controller manager tries to attach unmanaged azure disk (explicit blob) to my VMs with managed disks and then failed: no disks are attached eventually. 

Do I understand correctly that this PR will fix this problem and disks will be attached as managed disks to VMs with managed disks?

andrey-moor commented:
Is it going to be part of 1.7?

jdumars commented:
@andrey-moor we're going to break this work down into more consumable chunks and target 1.8, as well as cherry picks to prior versions.  There are a lot of moving parts here.

jdumars commented:
@nmakhotkin that is the intent, but please watch future PRs for how the work evolves.  I'll make sure any future work references this PR so it's easy to tie all the pieces together.

nmakhotkin commented:
@jdumars thanks!

andyzhangx commented:
@nmakhotkin In you case, you could only attach blob disks in your own created k8s cluster with this PR.
Only kubernetes cluster with managed os disk could attach managed data disks with this PR.
You could create k8s cluster with this acs template.
https://github.com/Azure/acs-engine/blob/master/examples/disks-managed/kubernetes-vmas.json

nmakhotkin commented:
@andyzhangx yes I already have the k8s cluster with VMs with managed OS disks. That is, without this PR k8s always tries to attach blob-based disk to VM and then fails.

andyzhangx commented:
@nmakhotkin  k8s cluster with managed os disk could only attach managed disk with this PR, I have tried, it works well. Below is the steps how to do this:
kubectl create -f https://raw.githubusercontent.com/khenidak/kubernetes-azure-managed-disks/master/claim/manageddisk/storageclass-managed-slow.json

kubectl create -f https://raw.githubusercontent.com/khenidak/kubernetes-azure-managed-disks/master/claim/manageddisk/pvc-on-managed-slow.json

kubectl create -f https://raw.githubusercontent.com/khenidak/kubernetes-azure-managed-disks/master/claim/manageddisk/pod-uses-managed-slow.json

nmakhotkin commented:
@andyzhangx thanks! 
It would be great if it was a part of 1.7 :(

andyzhangx commented on pkg/volume/azure_dd/common.go:
```diff
@@ -0,0 +1,322 @@
+	"path"
+	"regexp"
+	"strconv"
+	STRINGS "strings"
```
done

andyzhangx commented on pkg/volume/azure_dd/common.go:
vhd_util.go code has moved to common.go and vhd_uti_test.go code has moved to common_test.go

andyzhangx commented:
@rootfs  
Thans for the review, I am now taking over this PR.
As you could see, all CI tests passed.
Below are some comments of your comments, you could check other changes in the code, I have replied one by one after your comment in the code. Let me know if have missed sth, I would love to fix them all, thanks for your time.

**Refactoring:**
	• make sure existing methods in azure_dd.go are not lost
	[andy] Yes, they are all there, some existing methods has moved to azure_dd/mounter.go
	• Address mount utility related in issues, if any, in pkg/util/mount  
	[andy] TODO in next PR, discussed in PR
	• Keep vhd_util.go and its unit test
        [andy] actually code has moved to azure_dd/common.go, common_test.go

**Raw HTTP:**
	• If you continue raw HTTP, handle errors and malformed json.
	[andy] pls check the code, it's handled in func ExtractVMData, ExtractDiskData in auzre_util.go
	• Don't use hardcoded azure endpoints in raw HTTP, they don't work in every azure region.
	I am not an expert on Azure sdk, cc @ahmetb @colemickens
        [andy] fixed, use az.Environment.StorageEndpointSuffix

andyzhangx commented:
@rootfs 
Hi, do you have time to take a review? Thanks.

rootfs commented on pkg/volume/azure_dd/common.go:
```diff
@@ -0,0 +1,331 @@
+)
+
+const (
+	default_fstype               = "ext4"
```
camel case

andyzhangx replied:
fixed

rootfs commented on pkg/volume/azure_dd/common.go:
This issue is now discussed in #47158

rootfs commented on pkg/volume/azure_dd/common.go:
see #47158

rootfs commented on pkg/volume/azure_dd/provisioner.go:
```diff
@@ -0,0 +1,184 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
2017, and rename file name.

andyzhangx replied:
done, all new files renamed to azure_*.go

rootfs commented on pkg/volume/azure_dd/provisioner.go:
Address this.

rootfs commented on pkg/volume/azure_dd/provisioner.go:
Use case for storage account is for a multi tenant setup, providing separate storage account for each tenant.

rootfs commented on pkg/volume/azure_dd/provisioner.go:
let this for now, till #45345 is settled.

rootfs commented on pkg/volume/azure_dd/provisioner.go:
pvc -> storage class

rootfs commented on pkg/volume/azure_dd/provisioner.go:
fstype

rootfs commented on pkg/cloudprovider/providers/azure/managedDiskController.go:
```diff
@@ -0,0 +1,360 @@
+		return updateErr
+	}
+	// poll
+	// This is critical case, if this was a PVC, k8s will immediatly
```
immediately

andyzhangx replied:
done

rootfs commented on pkg/cloudprovider/providers/azure/controllerCommon.go:
```diff
@@ -0,0 +1,403 @@
+	return m
+}
+
+var mutex = &sync.Mutex{}
```
this won't work if the lun you get this time is not immediately consumed and filled by azure compute. You need a lock to protect findEmptyLun and attach disk.

andyzhangx replied:
fixed, i have used lock for findEmptyLun and append disk

rootfs reviewed (changes requested):
some quick feedback

jdumars commented:
@rootfs thanks for your continued reviews Huamin

andyzhangx commented on pkg/volume/azure_dd/provisioner.go:
fixed

andyzhangx commented on pkg/volume/azure_dd/provisioner.go:
skuname and storageAccountType has same meaning here, it's an alias:
Note that in older versions, sku name was called accountType.
type	SkuName { Standard_LRS, Standard_GRS, Standard_RAGRS, Standard_ZRS, Premium_LRS }

andyzhangx commented on pkg/volume/azure_dd/provisioner.go:
fixed.

rootfs commented on pkg/cloudprovider/providers/azure/azure_managedDiskController.go:
```diff
@@ -0,0 +1,366 @@
+	}
+
+	// lock for findEmptyLun and append disk
+	var mutex = &sync.Mutex{}
```
this lock goes out of range, it blocks attach to all nodes. Current lock is only for one instance, see https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/azure_dd/attacher.go#L98

khenidak replied:
Can you help me understand why we need this lock? i understand we are trying to prevent race conditions on getting free-lun-number that may result in getting the same lun. 

The following applies in context:
1. on the slim chance that two+ attach operations got the same lun, arm will apply one and reject the rest which will be retried by the controller. 
2. The current lock slows down the entire attach operations and unnecessary/partially serialize it in sequence.
3. The lock is in fact does protect against anything, two attach op can get the same when op A: lock, got lun#1, unlock, then op A's go routine de-schedule before running actual attach-call for op b: which will go through lock, get lun#1 (not yet reserved by op A), unlocked.

andyzhangx replied:
Hi kal, it's correct this lock logic is not reasonable.
I have combined AttachDisk process of blob-based disk and managed disk into one function, the lock process is following current existing logic. Just as rootfs commented, it's using https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/azure_dd/attacher.go#L98

rootfs commented on pkg/cloudprovider/providers/azure/azure_blobDiskController.go:
```diff
@@ -0,0 +1,1061 @@
+	// Azure in case of blob disks, does not maintain a list of vhd:attached-to:node
+	// The call  attach-to will fail after it was OK on the ARM VM endpoint
+	// possibly putting the entire VM in *failed* state
+	noLease, e := c.diskHasNoLease(diskURI)
```
This is yet another race condition: just because there is no lease on the disk doesn't mean somebody else wants to attach the disk before your attach request.

khenidak replied:
Actually no, the controller is attach/detach loop is sequential per PV. "Somebody else" in context refers to outside-k8s elements, between "no lease"  step and ARM placing its own lease. Those you can not control

andyzhangx replied:
this func is removed now, the lock process is following current existing logic

rootfs commented on pkg/cloudprovider/providers/azure/azure_controllerCommon.go:
```diff
@@ -0,0 +1,398 @@
+
+const (
+	aadTokenEndPointPath string = "%s/oauth2/token/"
+	apiversion           string = "2016-04-30-preview"
```
is this apiversion applicable for all? I glanced at azure sdk, there are several api versions for different services.

andyzhangx replied:
I have used sdk now, so these const are removed already.

rootfs commented:
@andyzhangx The race condition handling still needs work. 

@khenidak I am hugely concerned that this PR favors raw REST over Azure SDK. It is insane to let this go into production without sufficient test coverage. 

@jdumars I'll be out in the coming weeks. I'll handle lgtm over to @brendanburns

brendandburns commented:
I have PR that updates the go sdk so we should be able to move away from raw rest... It's already green and ready to go in.

--brendan

________________________________
From: Huamin Chen <notifications@github.com>
Sent: Thursday, June 22, 2017 6:14:01 AM
To: kubernetes/kubernetes
Cc: Brendan Burns; Mention
Subject: Re: [kubernetes/kubernetes] Azure PD (Managed/Blob) (#46360)


@andyzhangx<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fandyzhangx&data=02%7C01%7Cbburns%40microsoft.com%7Cb73783891ff94db6b17508d4b9708e3a%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636337340459240371&sdata=kt99G0UCcygCuTEnXcYiBXXMxo4shv%2FJ7cw%2B6auz2DY%3D&reserved=0> The race condition handling still needs work.

@khenidak<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fkhenidak&data=02%7C01%7Cbburns%40microsoft.com%7Cb73783891ff94db6b17508d4b9708e3a%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636337340459240371&sdata=Nylwv46Kx3PdI0X7WOGf%2FBLBmOWO7eXX4ZgkVV8L%2BEo%3D&reserved=0> I am hugely concerned that this PR favors raw REST over Azure SDK. It is insane to let this go into production without sufficient test coverage.

@jdumars<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fjdumars&data=02%7C01%7Cbburns%40microsoft.com%7Cb73783891ff94db6b17508d4b9708e3a%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636337340459240371&sdata=g6M5lFYUNJZYpVkHq1gR1YChZSa5B%2BKVpwWYbwQpJdQ%3D&reserved=0> I'll be out in the coming weeks. I'll handle lgtm over to @brendanburns<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fbrendanburns&data=02%7C01%7Cbburns%40microsoft.com%7Cb73783891ff94db6b17508d4b9708e3a%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636337340459240371&sdata=%2Bh0cXGI9ZrpETDVpi2N2C8kvWxQODRz9tS8f5sIM7F4%3D&reserved=0>

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fkubernetes%2Fkubernetes%2Fpull%2F46360%23issuecomment-310376010&data=02%7C01%7Cbburns%40microsoft.com%7Cb73783891ff94db6b17508d4b9708e3a%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636337340459240371&sdata=TwNHJqL01P2IIbT9BcX23iNdrtjqkCLiEhYICQoY4TQ%3D&reserved=0>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAFfDgtmDj_WQacitn3xIirwcGp_dkkZuks5sGmiZgaJpZM4NlXFu&data=02%7C01%7Cbburns%40microsoft.com%7Cb73783891ff94db6b17508d4b9708e3a%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636337340459240371&sdata=Fp666T62V2abjtGHS%2FArFbaOBUdZs40O9SOFcqwzOCg%3D&reserved=0>.

rootfs commented:
+1 @brendanburns

jdumars commented:
/sig azure

jdumars commented:
Adding SIG-Azure label as a stakeholder for tracking.

jdumars commented:
/retest

andyzhangx commented:
Hi @brendanburns, I am now working on changing raw REST code to using azure-go-sdk, and I found there managed disk API are supported starting from v8.0.0, currently k8s is using v7.0.1, so which azure-go-sdk version would you prefer to upgrade, below are the list:
v8.0.0-beta
v8.0.1-beta
v8.1.0-beta
v8.1.1-beta
v9.0.0-beta
v9.0.1-beta

andyzhangx commented:
Hi @brendanburns @rootfs @jdumars 
     I have changed to use azure sdk instead of REST api, combined the managed disk related code to the original code and fixed all comments raised by @rootfs.
For the compatibility issue, I have reserved the kind (dedicated) for compatability, which means below config still works: https://kubernetes.io/docs/concepts/storage/persistent-volumes/#azure-disk
     I have another PR shows the example and help about how to use this new PV plugin of azure managed/blob-based disk, you could find the help here:
   https://github.com/andyzhangx/kubernetes/tree/azure-disk-examples/examples/volumes/azure_disk/claim
     Let me know if you have any other comments, thanks.

andyzhangx commented:
/retest

jdumars commented:
@brendanburns @colemickens could you please have a look and see if it LGTY?

jdumars commented:
@brendandburns @colemickens @rootfs anyone able to give this a once-over so we can wrap it up?

brendandburns commented:
This LGTM and I'd like to get it in.

My SDK rev is just about to merge, let's wait for that and rebase this once it's in.

@rootfs any last comments before we merge and iterate?

rootfs commented:
@brendanburns just need a squash

brendandburns commented:
@andyzhangx can you rebase and squash, then I think we can merge...

andyzhangx committed:
Add Azure managed disk support

brendandburns commented:
@andyzhangx  Looks like there are some `govet` errors:

```
pkg/cloudprovider/providers/azure/azure_blobDiskController.go:612: arg err for printf verb %s of wrong type: <-chan error
```

And the underscore flag detector is also tripping:

```
I0713 13:20:30.188] Verifying hack/make-rules/../../hack/verify-flags-underscore.py
I0713 13:20:33.554] Found a flag declared with an _ but which is not explicitly listed as a valid flag name in hack/verify-flags/excluded-flags.txt
I0713 13:20:33.554] Are you certain this flag should not have been declared with an - instead?
I0713 13:20:33.554] Premium_LRS
```

I think that is a red-herring we should just add it to excluded-flags.txt...

karataliu commented:
/retest

karataliu commented:
@brendandburns  Thanks for reviewing. I've got the test fix commit https://github.com/karataliu/kubernetes/commit/3ed5502d438538f9c08fc2f93c28f646dc094622 verified, and it has been squashed into the final single commit  https://github.com/kubernetes/kubernetes/pull/46360/commits/677e593d864476d615d43006aada2007c80f0e16 here.

brendandburns commented:
/lgtm

brendandburns commented:
/approve no-issue

andyzhangx commented:
Hi @nmakhotkin , the code is already in 1.7.2, let me know if you have any question.

nmakhotkin commented:
@andyzhangx thanks! I'll try it.

jdumars commented:
Release Note:

Adds Kubernetes support for Azure Managed Disks, support for dedicated blob disks (1:1 to storage account) in addition to shared blob disks (n:1 to storage account), and automatically manages the underlying storage accounts. New storage accounts are created at 50% utilization. Max is 100 disks, 60 disks per storage account.   Due to the change in mounting paths, existing PDs need to be recreated as either PV or PVCs with the new plugin.
